[UnifiedPDF] PDF plugin should set background color on the PluginDocument

aprotyas opened the pull request:
#### 0d71b3d9d2b58719eceb9d45035e3d9c43939ea0
<pre>
[UnifiedPDF] PDF plugin should set background color on the PluginDocument
<a href="https://bugs.webkit.org/show_bug.cgi?id=286040">https://bugs.webkit.org/show_bug.cgi?id=286040</a>
<a href="https://rdar.apple.com/143008114">rdar://143008114</a>

Reviewed by Aditya Keerthi.

In 235203@main, we started overriding the plugin document&apos;s background
color in full main frame PDF plugins, allowing us to align with the
PDFLayerController layer backgrounds.

For UnifiedPDF, we want this same behavior, otherwise there are often
abrupt color transitions on orientation changes (among other things) on
iOS, since there is now a mismatch between the plugin&apos;s background layer
color (rgb(127, 127, 127)) and the plugin document&apos;s background color
(rgb(217, 224, 233)).

To facilitate this, we introduce PDFPluginBase::pluginBackgroundColor(),
which just ends up consulting [NS|UI]Color.grayColor, and adopt that
across the various codepaths where we assign background colors in the
PDF plugin.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::PDFPlugin):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::PDFPluginBase):
(WebKit::PDFPluginBase::pluginBackgroundColor):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::ensureLayers):
(WebKit::UnifiedPDFPlugin::paint):

Canonical link: <a href="https://commits.webkit.org/289013@main">https://commits.webkit.org/289013@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1925f73fe81a96ff10ff57ae06c231d25ff168d6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84985 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4714 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39381 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90134 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36041 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4805 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12690 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66138 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23954 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88030 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3677 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77235 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46404 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3557 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35114 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32273 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91507 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12329 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9014 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74621 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12558 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73046 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73741 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18261 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18132 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16584 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4350 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12277 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12113 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15609 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13858 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -539,12 +535,6 @@ - (PDFPage *)pageNearestPoint:(NSPoint)point currentPage:(PDFPage *)currentPage;
 
-    bool isFullFrame = isFullFramePlugin();
-    if (isFullFrame) {
-        // FIXME: <rdar://problem/75332948> get the background color from PDFKit instead of hardcoding it
```
Should we keep the FIXME around, and move it to `PDFPluginBase::pluginBackgroundColor`?

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -1460,6 +1468,11 @@ static void verboseLog(PDFIncrementalLoader* incrementalLoader, uint64_t streame
 
+Color PDFPluginBase::pluginBackgroundColor()
+{
+    return roundAndClampToSRGBALossy([CocoaColor grayColor].CGColor);
```
Maybe we should store this in a `static NeverDestroyed`. It'll get called every time the PDF is painted right?

aprotyas committed:
[UnifiedPDF] PDF plugin should set background color on the PluginDocument
https://bugs.webkit.org/show_bug.cgi?id=286040
rdar://143008114

Reviewed by Aditya Keerthi.

In 235203@main, we started overriding the plugin document's background
color in full main frame PDF plugins, allowing us to align with the
PDFLayerController layer backgrounds.

For UnifiedPDF, we want this same behavior, otherwise there are often
abrupt color transitions on orientation changes (among other things) on
iOS, since there is now a mismatch between the plugin's background layer
color (rgb(127, 127, 127)) and the plugin document's background color
(rgb(217, 224, 233)).

To facilitate this, we introduce PDFPluginBase::pluginBackgroundColor(),
which just ends up consulting [NS|UI]Color.grayColor, and adopt that
across the various codepaths where we assign background colors in the
PDF plugin.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::PDFPlugin):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::PDFPluginBase):
(WebKit::PDFPluginBase::pluginBackgroundColor):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::ensureLayers):
(WebKit::UnifiedPDFPlugin::paint):

Canonical link: https://commits.webkit.org/289013@main
